Add --reboot-exit-code parameter to "install" command.

hacpa opened the pull request:
This parameter will cause the "mender install" command to return an
exit code of 4 if a manual reboot is required after installation.

Changelog: Title
Signed-off-by: Pascal Hache <hacpa@touchtunes.com>


# External Contributor Checklist

🚨 Please review the [guidelines for contributing](../CONTRIBUTING.md) to this repository.

- [x] Make sure that all commits have a [`Changelog`](../CONTRIBUTING.md#changelog-tags) tag. If nothing should be added to the Changelog, add a `Changelog: None` tag. If there is a change, add `Changelog: Commit`, or `Changelog: Title`, depending on what should be included in the changelog.

- [x] Make sure that all commits are signed with [`git --signoff`](https://git-scm.com/book/en/v2/Git-Tools-Signing-Your-Work). Also note that the signoff author must match the author of the commit.

### Description

Please describe your pull request.

Thank you!

oleorhagen commented on app/standalone.go:
```diff
@@ -237,6 +242,9 @@ func doStandaloneInstallStates(art io.ReadCloser, key []byte,
 
 	if rebootNeeded {
 		fmt.Println("At least one payload requested a reboot of the device it updated.")
+		if rebootExitCode == true {
```
Nitpick:
```suggestion
		if rebootExitCode {
```

hacpa replied:
Done.

oleorhagen commented on cli/cli.go:
```diff
@@ -172,6 +173,14 @@ func SetupCLI(args []string) error {
+				&cli.BoolFlag{
+					Name:        "reboot-exit-code",
+					Destination: &runOptions.rebootExitCode,
+					Usage: "Set exit code to 4 if a manual reboot " +
```
Suggestion:
`Return exit code 4 If a manual reboot is required after the Artifact installation.`

oleorhagen replied:
Notice the Uppercase `Artifact` especially.

hacpa replied:
Done.

oleorhagen commented on app/standalone.go:
```diff
@@ -44,7 +48,7 @@ type standaloneData struct {
 // This will be run manually from command line ONLY
 func DoStandaloneInstall(device *dev.DeviceManager, updateURI string,
```
Suggestion:
It could maybe be time to factor this call out into accepting a struct, which has the standard values set already.

So something like:
```
type doStandaloneInstallArgs struct {
    rebootExitCode bool
   ...
}
```
If so, add this in a separate commit, so that the other reviewers can easily chime in their opinions without creating a lot more work for you :)

hacpa replied:
I feel like this rework is outside the scope of my change. I'll leave this one to the pros, if you don't mind. :)

oleorhagen replied:
That is just fine :)

oleorhagen reviewed (commented):
Real nice!

Just some minor stuff.

1. I don't see an actual test case for this. Just add a simple unit test please.
2. Nitpicks

kacf reviewed (changes requested):
Looks good @hacpa!

My only comment apart from what @oleorhagen already said, is that this should be targeted to the `master` branch. We don't add features in stable branches. Don't worry, the cherry-pick should be trivial, I don't think this part has changed much.

hacpa commented:
@oleorhagen 
1. I can't actually test the app. The test of the "app" package takes forever and panics after 10 minutes of running at 100% CPU. You may have noticed that my initial push failed the tests and that I force-pushed another commit to fix that. I can try making a test case, but I can't guarantee that it will actually test anything.
2. Just a question in regards to your process. Should I make new individual commits for each suggestion (I see I can just press a button to automatically create a commit for each suggestion), I should I bundle everything into an amend and force-push it? I don't have a problem with applying the trivial suggestions, but I don't want to cause you merge problems either.

@kacf I can't actually compile the master branch (some unmet dependencies that are not mentioned in the readme). I'm not sure how to change the target to master either. Do I just cherry-pick to another branch based on master and create a new PR?

lluiscampos commented:
> 
>     2. Just a question in regards to your process. Should I make new individual commits for each suggestion (I see I can just press a button to automatically create a commit for each suggestion), I should I bundle everything into an amend and force-push it? I don't have a problem with applying the trivial suggestions, but I don't want to cause you merge problems either.
> 

Please bundle everything, amend, force-push. We prioritize for a sane Git history :)

> 
> @kacf I can't actually compile the master branch (some unmet dependencies that are not mentioned in the readme). 

Maybe OpenSSL dev library? Can you copy us the error you are getting?

> I'm not sure how to change the target to master either. Do I just cherry-pick to another branch based on master and create a new PR?

I tried to do this myself but wasn't a good idea. You can do two things: 1) rebase this branch on top of latest master, and then click "Edit" above to change the base of the PR, or 2) close this PR, create new branch from master, cherry-pick this commit, launch a new PR.

lluiscampos commented:
> Hello smile_cat I created a pipeline for you here: [Pipeline-188401241](https://gitlab.com/Northern.tech/Mender/mender-qa/-/pipelines/188401241)
> Build Configuration Matrix

Stopping the build as this PR still needs rework.

hacpa commented:
> 
> 
> > ```
> > 2. Just a question in regards to your process. Should I make new individual commits for each suggestion (I see I can just press a button to automatically create a commit for each suggestion), I should I bundle everything into an amend and force-push it? I don't have a problem with applying the trivial suggestions, but I don't want to cause you merge problems either.
> > ```
> 
> Please bundle everything, amend, force-push. We prioritize for a sane Git history :)
> 
> > @kacf I can't actually compile the master branch (some unmet dependencies that are not mentioned in the readme).
> 
> Maybe OpenSSL dev library? Can you copy us the error you are getting?
> 
> > I'm not sure how to change the target to master either. Do I just cherry-pick to another branch based on master and create a new PR?
> 
> I tried to do this myself but wasn't a good idea. You can do two things: 1) rebase this branch on top of latest master, and then click "Edit" above to change the base of the PR, or 2) close this PR, create new branch from master, cherry-pick this commit, launch a new PR.

libssl.pc and libcrypto.pc are missing. Perhaps the readme should be updated?

I actually need this branch to remain intact and on top of 2.3.x (I am using it directly in my project while I wait for the changes to be integrated in the main 2.3.x), so I will choose option 2 - new PR based on master with new branch - once everyone is satisfied with what's in here. New changes coming soon.

hacpa commented:
Re-pushed. While I can't run the full tests on the "app" package, I was able to run the individual test case I created (copied the successful test case just above it).

Please let me know if my change is now acceptable. Then as I said above, I will close this PR and create a new one of top of master.

oleorhagen reviewed (approved):
Looks good to me now. Good job!

hacpa commented:
There, ended up choosing option 1 proposed by @oleorhagen . This PR is now based on master. I created another branch on my fork to keep the change where I want it (temporarily), so the branch "reboot_return_code" may be deleted when this is merged.

oleorhagen commented:
Rebased on top of `master` for a test-run here: https://github.com/mendersoftware/mender/pull/603

oleorhagen commented:
@hacpa I've fixed the test errors, and rebased this on top of master in #603 . If this is okay with you, I will close this PR, and merge the other one?

hacpa commented:
> 
> 
> @hacpa I've fixed the test errors, and rebased this on top of master in #603 . If this is okay with you, I will close this PR, and merge the other one?

Looks good to me!

oleorhagen commented:
Closed in favour of #603
